### ISSUE TITLE
Wayland support?

### ISSUE THREAD

**Issue by deralmas** [zserge/fenster#6]
Hi! This looks like a really nice project! I wonder if it would make sense to support Wayland here. I could give a hand if you're interested.

**Comment by diamondburned** [comment]
Given Fenster's API, it'll be quite tough to have HiDPI support for Wayland. The main issue is that the client expects a fixed viewport with an image buffer of the same size, so we can't let the user have a larger image buffer for scaling. If we add a new API, the user will have to worry a lot more when blitting images onto Fenster.

I think Wayland support can still be done with something like [shared buffers](https://wayland-book.com/surfaces/shared-memory.html) though. It'll just be a lot more code, since Wayland's client API is a lot more code than X's higher-level libraries. We'll also need something like [libdecoration](https://gitlab.gnome.org/jadahl/libdecoration) added on top, probably.

**Comment by diamondburned** [comment]
> AFAICT there's no X11/Windows/MacOS HiDPI support either, so I think that there should be no problem.

That's true, however being a Wayland app, I think it should at least have some ability to render without being blurry on modern computers.

Perhaps just using nearest-neighbor integer scaling would still yield better results than letting the compositor scale up.

> I'm not really fond of libdecoration and I'd really like to avoid such an extra dependency if possible. Perhaps we could make some "inline" decoration drawing routine?

That's fair. Given the simplicity of this library, I think we can get away with rendering a blank deco only once with basic buttons.

**Comment by diamondburned** [comment]
> Is the filter used for HiDPI buffers defined somewhere in the spec?

I don't think so, but afaik, most compositors use whatever OGL has built-in, which is bilinear filtering.
